modules/imgproc/src/color_hsv.simd.hpp

```diff
@@ -39,36 +43,58 @@ struct RGB2HSV_b
     : srccn(_srccn), blueIdx(_blueIdx), hrange(_hrange)
     {
         CV_Assert( hrange == 180 || hrange == 256 );
+        initTables_();
     }
 
-    void operator()(const uchar* src, uchar* dst, int n) const
+    void initTables_()
     {
-        CV_INSTRUMENT_REGION();
-
-        int i, bidx = blueIdx, scn = srccn;
         const int hsv_shift = 12;
 
         static int sdiv_table[256];
         static int hdiv_table180[256];
         static int hdiv_table256[256];
+#ifdef CV_CXX11
+        static std::atomic<bool> initialized(false);
+#else
         static volatile bool initialized = false;
+#endif
 
-        int hr = hrange;
-        const int* hdiv_table = hr == 180 ? hdiv_table180 : hdiv_table256;
+        hdiv_table_ = hrange == 180 ? hdiv_table180 : hdiv_table256;
+        sdiv_table_ = sdiv_table;
 
-        if( !initialized )
+#ifdef CV_CXX11
+        if (!initialized.load(std::memory_order_acquire))
+#else
+        if (!initialized)
+#endif
         {
             sdiv_table[0] = hdiv_table180[0] = hdiv_table256[0] = 0;
```

Comment:
I think it's a data race to write the values from different threads to one memory location, even if these values are the same. It might be better to create a struct with those tables and perform calculations in constructor. Then just create static variable of that type and c++11 will guarantee that the initialization is thread-safe.

Reply:
replaced by #22061